perfkitbenchmarker/digitalocean/digitalocean_virtual_machine.py

```diff
@@ -0,0 +1,211 @@
+  - name: {0}
+    ssh-authorized-keys:
+      - {1}
+    sudo: ['ALL=(ALL) NOPASSWD:ALL']
+    groups: sudo
+    shell: /bin/bash
+'''
+
+class DigitaloceanVirtualMachine(virtual_machine.BaseVirtualMachine):
+  """Object representing a DigitalOcean Virtual Machine (Droplet)."""
+
+  def __init__(self, vm_spec):
+    """Initialize a DigitalOcean virtual machine.
+
+    Args:
+      vm_spec: virtual_machine.BaseVirtualMachineSpec object of the vm.
+    """
+    super(DigitaloceanVirtualMachine, self).__init__(vm_spec)
+    self.droplet_id = None
+    self.max_local_disks = 1
+    self.local_disk_counter = 0
+
+  def _CreateDependencies(self):
+    """Create VM dependencies."""
+    pass
+
+  def _DeleteDependencies(self):
+    """Delete VM dependencies."""
+    pass
+
+  def _Create(self):
+    """Create a DigitalOcean VM instance (droplet)."""
+    super(DigitaloceanVirtualMachine, self)._Create()
+    with open(self.ssh_public_key) as f:
+      public_key = f.read().rstrip('\n')
+
+    stdout, ret = util.RunCurlCommand(
+        'POST', 'droplets', {
+            'name': self.name,
+            'region': self.zone,
+            'size': self.machine_type,
+            'image': self.image,
+            'backups': False,
+            'ipv6': False,
+            'ssh_keys': [],
+            'user_data': CLOUD_CONFIG_TEMPLATE.format(
+                self.user_name, public_key)
+            })
+    if ret != 0:
+      raise errors.Error('Creation failed: %s' % stdout)
+    response = json.loads(stdout)
+    self.droplet_id = response['droplet']['id']
+
+    # The freshly created droplet will be in a locked and unusable
+    # state for a while, and it cannot be deleted or modified in
+    # this state. Wait for the action to finish and check the
+    # reported result.
+    if not self._GetActionResult(response['links']['actions'][0]['id']):
+      raise errors.Resource.RetryableCreationError('Creation failed.')
+
+  def _GetActionResult(self, action_id, wait_seconds=10, max_tries=30):
+    """Wait until a VM action completes."""
+    for _ in xrange(max_tries):
+      time.sleep(wait_seconds)
+      stdout, ret = util.RunCurlCommand('GET', 'actions/%s' % action_id)
+      if ret != 0:
+        raise errors.Error('Unexpected action failure: %s' % stdout)
+      response = json.loads(stdout)
+      status = response['action']['status']
+      if status == 'completed':
+        return True
+      elif status == 'errored':
+        return False
+    return False  # Timeout, treat as failure.
+
+  @vm_util.Retry()
+  def _PostCreate(self):
+    """Get the instance's data."""
+    stdout, _ = util.RunCurlCommand(
+        'GET', 'droplets/%s' % self.droplet_id)
+    response = json.loads(stdout)['droplet']
+    network_interface = response['networks']['v4'][0]
+    # Not currently using private addresses. Do we need to?
+    self.internal_ip = network_interface['ip_address']
+    self.ip_address = network_interface['ip_address']
+
+  @vm_util.Retry()
```

Comment:
Do you need the retry around this function? If deleting a VM that doesn't exist any more is an error (non zero retcode), you'll get stuck in a loop until the retry timeout.